### PR TITLE
bulk submission for parsl executor

### PR DESCRIPTION
**Pull request by andre-merzky**
This PR improves submission throughput by collecting task bulks on the PARSL layer before pushing to RP.  The max task rate I see is still not great though, about 20/sec. (improved from 5/sec though).

### CODE REVIEW

**Comment by andre-merzky** [comment]
We follow up with that once Aymen's experiments are done

**Comment by andre-merzky** [comment]
TODO: make bulk size configurable

**Comment by AymenFJA** [comment]
To be compatible with other use cases that require `task stream submission mode` like `Colmena`, `bulk mode` is now optional.

File: src/radical/pilot/agent/executing/parsl_rp.py [hunk 1]
```diff
@@ -351,16 +352,9 @@ def submit(self, func, *args, **kwargs):
...
-        # FIXME: use uid generator
-        task_id = 'task.parsl.%06d' % self._task_counter
-
-      # # ----------------------------------------------------------------------
-      # # test code: this is the fastest possible executor implementation
-      # self.future_tasks[task_id] = Future()
-      # self.future_tasks[task_id].set_result(3)
-      # return self.future_tasks[task_id]
-      # # ----------------------------------------------------------------------
+        self._task_counter = ru.generate_id('task.%(item_counter)06d',
```
Comment (andre-merzky) [inline src/radical/pilot/agent/executing/parsl_rp.py:355]: Why do we need the `self._task_counter variable?

File: examples/parsl/map_reduce.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 config = Config(
          executors=[RADICALExecutor(
                         label        = 'RADICALExecutor',
+                        bulk_mode    = False,
```
Comment (andre-merzky) [inline examples/parsl/map_reduce.py:15]: Should we turn on the bulk mode for this example, or is there any drawback?

**Comment by andre-merzky** [comment]
The optional bulk mode makes sense IMHO.

File: src/radical/pilot/agent/executing/parsl_rp.py (hunk 1 above)
Comment (AymenFJA) [reply src/radical/pilot/agent/executing/parsl_rp.py:355]: For two reasons: 

1. Maintain consistency in terms of task numbers between RP and Parsl (we create a future that the DFK needs to look for when fetching the task info if it's finished/pending/running).
2. I followed the class code template provided by `HTEX` for `RP-Parsl` executor which used `self._task_counter`.

File: examples/parsl/map_reduce.py (hunk 2 above)
Comment (AymenFJA) [reply examples/parsl/map_reduce.py:15]: @andre-merzky sure, that makes sense.

File: src/radical/pilot/agent/executing/parsl_rp.py (hunk 1 above)
Comment (AymenFJA) [reply src/radical/pilot/agent/executing/parsl_rp.py]: Ohh, I got it and makes sense. I addressed your proposal in the latest commit.

File: examples/parsl/map_reduce.py (hunk 2 above)
Comment (AymenFJA) [reply examples/parsl/map_reduce.py]: Addressed.

**Comment by andre-merzky** [comment]
I can't approve the PR as I opened it, but that looks good to merge from my perspective.

**Review by mtitov — APPROVED** [review]
LGTM!

**Comment by AymenFJA** [comment]
Thank you all. I will merge now!

**Pull request merged**